core/observe.c

```diff
@@ -950,35 +980,57 @@ static void prv_obsRequestCallback(lwm2m_context_t * contextP,
 
             // give the user chance to free previous observation userData
             // indicator: COAP_202_DELETED and (Length ==0)
-            observationP->callback(observationP->clientP->internalID,
-                    &observationP->uri,
-                    COAP_202_DELETED,
-                    0, NULL, 0,
-                    observationP->userData);
+            prv_applyObservationRequestCallback(observationData, COAP_202_DELETED, block_num, block_size, block_more, 0, NULL, 0);
         }
 
         observationP->id = observationData->id;
         observationP->clientP = clientP;
 
         observationP->callback = observationData->callback;
+        observationP->blockCallback = observationData->blockCallback;
         observationP->userData = observationData->userData;
         observationP->status = STATE_REGISTERED;
         memcpy(&observationP->uri, uriP, sizeof(lwm2m_uri_t));
 
         observationP->clientP->observationList = (lwm2m_observation_t *)LWM2M_LIST_ADD(observationP->clientP->observationList, observationP);
 
-        observationData->callback(observationData->client,
-                &observationData->uri,
-                0,
-                packet->content_type, packet->payload, packet->payload_len,
-                observationData->userData);
+        uint32_t block_num = 0;
+        uint16_t block_size = 0;
+        uint8_t block_more = 0;
+        coap_get_header_block(message, &block_num, &block_more, &block_size, NULL);
+        prv_applyObservationRequestCallback(observationData, 0, block_num, block_size, block_more, utils_convertMediaType(packet->content_type), packet->payload, packet->payload_len);
     }
 
 end:
-    lwm2m_free(observationData);
+    transaction_free_userData(contextP, transacP);
+}
+
+
+static
+void prv_applyCancelRequestCallback(cancellation_data_t * cancelP, int status, int block_num, int block_size, bool block_more, lwm2m_media_type_t format, uint8_t * data, int dataLength)
+{
+    if (cancelP->blockCallbackP != NULL)
+    {
+        cancelP->blockCallbackP(cancelP->client,
+                &cancelP->uri,
+                status,  //?
+                block_num, block_size, block_more,
+                format,data, dataLength,
+                cancelP->userDataP);
+    }
+    else if (cancelP->callbackP != NULL)
+    {
+        cancelP->callbackP(cancelP->client,
```

Comment:
Calling callbacks without checking arguments is not nice, who knows what the callback does...

Reply:
I think we had similar discussions in other cases and the argument beeing that in order to keep a minimal footprint we need to trust calling functions and callback to not do anything vile.